Allow entire test suite to run with multiple compaction groups

raphaelsc opened the pull request:
New test/lib/scylla_test_case.hh, introduced in "tests: Add command line options for Scylla unit tests",
allows extension of the command line options provided by Seastar testing framework.
It allows all boost tests to process additional options without changing a single line of code.

Patch "test: Add x-log2-compaction-groups to Scylla test command line options" builds on that, allowing
all test cases to run with N compaction groups. Again, without changing a line of code in the tests.

Now all you have to do is:
./build/dev/test/boost/sstable_compaction_test -- --smp 1 --x-log2-compaction-groups 1
./test.py --mode=dev --x-log2-compaction-groups 1 --verbose

And it will run the test cases with as many groups as you wish.

./test.py passes successfully with parameter --x-log2-compaction-groups 1.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3567/

denesb commented on test/cql-pytest/test_bloom_filter.py:
```diff
@@ -28,7 +28,7 @@ def test_bloom_filter(scylla_only, cql, test_keyspace, N, M, fp_chance):
           "bloom_filter_false_positives", table)
         ratio = fp / M
-        assert ratio >= fp_chance * 0.7 and ratio <= fp_chance * 1.15
+        assert fp > 0 and ratio <= fp_chance * 1.15
```
Can we instead make sure all test data is in a single group? If that is too difficult, it might not be worth it.

avikivity replied:
I think this relaxes the test too much.

I don't understand why multiple compaction groups make the bloom filter more efficient. I guess they do, if there is a too-low number of partitions.

raphaelsc replied:
```
pkeys		ssts		pks:sst		data		filter		data:filter
100		1		100		4300		136		31
100		10		10		4300		240		17
100		100		1		4300		1600		2
1000		1		1000		43872		1264		34
1000		10		100		43000		1360		31
1000		100		10		43000		2400		17
10000		1		10000		439872		12512		35
10000		10		1000		438720		12640		34
10000		100		100		430000		13600		31
100000		1		100000		4483488		125016		35
100000		10		10000		4398720		125120		35
100000		100		1000		4387200		126400		34
```

data above shows that with < 1k pkeys, data:filter ratio increases significantly as number of sstables increases.  with 10k keys, there's barely any difference. with 100k keys, even less.

we have two options, resample bloom filters to meet a data:filter ratio goal, but that's a long term effort.

alternatively, I can just increase number of keys to 10k, with the test time going from 1.17s to 4.39s. sounds good to you?

raphaelsc replied:
An idea: run the test with 500 pkeys as today. On failure, switch to slow mode which will try again with 10k-100k pkeys. Of course, we're embedding implementation detail into test, which isn't the best thing to do, but that's a reasonable assumption that bloom filter should increase in efficiency if the underlying database stores the keys inserted into multiple bloom filters, one per sstable. So if first line of defense fails, we go into slow mode, but for regular runs, fast mode happens.

denesb commented on replica/table.cc:
```diff
@@ -1414,7 +1414,11 @@ table::table(schema_ptr schema, config config, db::commitlog* cl, compaction_man
     , _x_log2_compaction_groups(_config.x_log2_compaction_groups)
+#else
+    , _x_log2_compaction_groups(FORCE_X_LOG2_COMPACTION_GROUPS)
+#endif
```
I think it would be better if we just changed the default for the above config item based on the macro.
See existing precendents: https://github.com/scylladb/scylladb/blob/d9269abf5b50c8f2a0c45334f6169691ef06c9a5/db/config.cc#L234-L244

avikivity commented:
> Directive FORCE_X_LOG2_COMPACTION_GROUPS can be used to allow all binaries to forcefully use 2^FORCE_X_LOG2_COMPACTION_GROUPS compaction groups, allowing the entire suite to stress multiple groups.
> 
> e.g.: ./configure.py --mode=dev --cflags='-DFORCE_X_LOG2_COMPACTION_GROUPS=3'; ninja -j28 dev-build;
> 

Let's use a runtime flag instead of forcing recompilation. After all it's runtime enabled.

> will compile all binaries with 8 groups.
> 
> Series also include fixes for tests which were failing with multiple groups.

avikivity commented:
I thought we had a runtime flag for it.

raphaelsc commented:
> I thought we had a runtime flag for it.

@avikivity I am not aware of any such a mechanism. Boost tests themselves don't parse command line args, so we're limited by the options defined in seastar testing infrastructure. So I went for a new test option, but of course it's limited to tests which launch a scylla instance. 

/CC @denesb

raphaelsc commented:
> I thought we had a runtime flag for it.

For boost tests, we can employ a condition variable which class table_for_tests (wraps table for tests) can check for running with multiple groups.
For tests running against a Scylla instance, it makes sense to just forward the option which is processed in the startup.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3978/

raphaelsc commented:
v3:

New test/lib/scylla_test_case.hh, introduced in "tests: Add command line options for Scylla unit tests"
allows extension of the command line options provided by Seastar testing framework.
It allows all boost tests to process additional options *without* changing a single line of code.

Patch "test: Add x-log2-compaction-groups to Scylla test command line options" builds on that, allowing
all test cases to run with N compaction groups. Again, *without* changing a line of code in the tests.

Now all you have to do is:
./build/dev/test/boost/sstable_compaction_test -- --smp 1 --x-log2-compaction-groups 1
./test.py --mode=dev --x-log2-compaction-groups 1 --verbose

And it will run the test cases with as many groups as you wish.

./test.py passes successfully with parameter --x-log2-compaction-groups 1.

raphaelsc commented:
/cc @avikivity

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4009/

avikivity commented on test/boost/memtable_test.cc:
```diff
@@ -984,9 +984,9 @@ SEASTAR_TEST_CASE(failed_flush_prevents_writes) {
-        utils::get_local_injector().enable("table_seal_active_memtable_add_memtable", true /* oneshot */);
-        utils::get_local_injector().enable("table_seal_active_memtable_start_op", true /* oneshot */);
-        utils::get_local_injector().enable("table_seal_active_memtable_try_flush", true /* oneshot */);
+        utils::get_local_injector().enable("table_seal_active_memtable_add_memtable");
```
1. it may make it "too easy" for the test to pass if we inject many errors, reducing test selectivity
2. it may pollute things further on (but I see you remembered to disable the injections)

Can we change the condition so it works even if just one memtable fails to flush?

avikivity commented on test/lib/scylla_test_case.hh:
```diff
@@ -0,0 +1,37 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
```
Please use our regular SPDX AGPL declaration. Otherwise, pretty nice.

raphaelsc replied:
Will change. thanks.

avikivity commented on test/lib/scylla_tests_cmdline_options.hh:
```diff
@@ -0,0 +1,36 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
```
Here too.

avikivity commented:
Nice not to have to recompile everything!

raphaelsc commented:
v4:

- reworked fixes for test_bloom_filter and memtable_test.
- adjusted licenses.

"./test.py --mode=dev --x-log2-compaction-groups 2" succeeded.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4035/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4034/

denesb commented on test/lib/test_services.cc:
```diff
@@ -156,3 +159,68 @@ test_env::impl::impl(test_env_config cfg)
+
+static std::pair<int, char**> rebuild_arg_list_without(int argc, char** argv, const char* filter_out, bool exclude_positional_arg = false) {
+    int new_argc = 0;
+    char** new_argv = (char**) malloc(argc * sizeof(char*));
```
You don't need to reallocate the arg list, you can just swap out unwanted items to the end, then --argc accordingly.

raphaelsc replied:
That's indeed an alternative, but I found it more convenient to do any changes in a copy as the cmd separator '--' has to be restored to its original position. Also, test.py time is dominated by tests themselves, so rebuilding arg list for each test file doesn't make any real difference. Checked it verifying the elapsed time before and after. If each rebuilding takes ~5us (takes even less), then total time is 2.6k tests * ~5us = ~13ms. Not significant even if we grow number of tests to 100k, which would result in 0.5s overhead.

denesb replied:
With the current method, any exception will result in an ASAN leak report, possibly causing whoever has to investigate it, to start on the wrong tracks.
At the very least use `std::unique_ptr<>` to keep the allocated args alive.
The advantage of shifting arguments around is that you don't have to deal with lifetime. As long as the new arg list is shorter than the old one, you can rearrange the args any way you want.

raphaelsc replied:
The arg list is automatically freed on the class destructor (see ~scylla_tests_cmdline_options_processor()).  So ASAN wise, we're good. The disadvantage with shifting arguments is that I have to undo changes, which is both tricky and error prone. The current approach is more convienient and has no significant overhead.

denesb replied:
I see an explicit call to `free_arg_list(new_argc, new_argv);` below. So is the arglist freed twice?

raphaelsc replied:
The argument list feeded into the test is always stored in scylla_tests_cmdline_options_processor::_new_argv (see patch introducing x-log2-... option). The one destroyed explicitly is temporary and only for processing the arg list without the '--' separator. I am thinking about using defer() for automatically destroying it. Sounds good to you?

denesb replied:
Yes, that sounds good, although `std::unique_ptr<>` would be better, but it is indeed more clunky to use here.

raphaelsc commented:
@scylladb/scylla-maint https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4034/ succeeded. https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4035/ (same commit hash as 4034) failed for unrelated reason. Anyway, I re-kicked CI just in case.

raphaelsc commented:
v5:
use defer() for automatically destroying temporary arg list.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4063/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4070/

denesb commented:
Please rebase, there are conflicts. It is possible that the conflicting patch is only next, as GH is not complaining about conflicts here in the PR.

raphaelsc commented:
v6: rebased (resolved minor conflict in patch "replica: table: Add debug log for number of compaction groups")

raphaelsc commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4128/

unrelated failure on raft test. @denesb you can only merge it into next with a successful CI right?

denesb commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4128/
> 
> unrelated failure on raft test. @denesb you can only merge it into next with a successful CI right?

Yes, even unrelated failures can hide genuine ones. So we need a successful CI run to merge.

denesb commented:
Re-kicked CI.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4142/

raphaelsc commented:
> CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4142/

Finally :-) 

/CC @denesb
